lib/FrontendTool/FrontendTool.cpp

```diff
@@ -262,6 +264,318 @@ template <> struct ObjectTraits<LoadedModuleTraceFormat> {
+  module->getImportedModules(importList, importFilter);
+
+  for (ModuleDecl::ImportedModule &import : importList)
+    imports.insert(import.importedModule);
+}
+
+namespace {
+/// Helper type for computing (approximate) information about ABI-dependencies.
+///
+/// This misses out on details such as typealiases and more.
+/// See the "isImportedDirectly" field above for more details.
+class ABIDependencyEvaluator {
+  /// Map of ABIs exported by a particular module, excluding itself.
+  ///
+  /// For example, consider (primed letters represent Clang modules):
+  /// \code
+  /// - A is @_exported-imported by B
+  /// - B is #imported by C' (via a compiler-generated umbrella header)
+  /// - C' is @_exported-imported by C (Swift overlay)
+  /// - D' is #imported by E'
+  /// - D' is @_exported-imported by D (Swift overlay)
+  /// - E' is @_exported-imported by E (Swift overlay)
+  /// \endcode
+  ///
+  /// Then the \c abiExportMap will be
+  /// \code
+  /// { A: {}, B: {A}, C: {B}, C': {B}, D: {}, D': {}, E: {D}, E': {D'} }
+  /// \endcode
+  ///
+  /// \b WARNING: Use \c reexposeImportedABI instead of inserting directly.
+  llvm::DenseMap<ModuleDecl *, llvm::DenseSet<ModuleDecl *>> abiExportMap;
+
+  /// Stack for depth-first traversal.
+  SmallVector<ModuleDecl *, 32> searchStack;
+
+  llvm::DenseSet<ModuleDecl *> visited;
+
+  /// Helper function to handle invariant violations as crashes in debug mode.
+  void crashOnInvariantViolation(
+    llvm::function_ref<void (llvm::raw_string_ostream &)> f) const;
+
+  /// Computes the ABI exports for \p importedModule and adds them to
+  /// \p module's ABI exports.
+  ///
+  /// If \p includeImportedModule is true, also adds \p importedModule to
+  /// \p module's ABI exports.
+  ///
+  /// Correct way to add entries to \c abiExportMap.
+  void reexposeImportedABI(ModuleDecl *module, ModuleDecl *importedModule,
+                           bool includeImportedModule = true);
+
+  /// Recursive step in computing ABI dependencies.
+  ///
+  /// Use this method instead of using the \c forClangModule/\c forSwiftModule
+  /// methods.
+  void computeABIDependenciesForModule(ModuleDecl *module);
+  void computeABIDependenciesForSwiftModule(ModuleDecl *module);
+  void computeABIDependenciesForClangModule(ModuleDecl *module);
+
+  static void printModule(const ModuleDecl *module, llvm::raw_ostream &os);
+
+  template<typename SetLike>
+  static void printModuleSet(const SetLike &set, llvm::raw_ostream &os);
+
+public:
+  ABIDependencyEvaluator() = default;
+  ABIDependencyEvaluator(const ABIDependencyEvaluator &) = delete;
+  ABIDependencyEvaluator(ABIDependencyEvaluator &&) = default;
+
+  void getABIDependenciesForSwiftModule(
+    ModuleDecl *module, SmallPtrSetImpl<ModuleDecl *> &abiDependencies);
+
+  void printABIExportMap(llvm::raw_ostream &os) const;
+};
+} // end anonymous namespace
+
+// See [Note: Bailing-vs-crashing-in-trace-emission].
+// TODO: Use PrettyStackTrace instead?
```

Comment:
It'd be a good idea to use plain assert + PrettyStackTrace for context. That way the contextual info will also be available for unanticipated crashes.

Reply:
I can do that in a follow-up PR. There is another function here which needs to use PrettyStackTrace instead of erroring out manually, it would be good to handle the two of them in the same PR.